### ISSUE TITLE
Harden Grommet Site static site build and deploy

### ISSUE THREAD

**Issue by halocline** [grommet/grommet-site#200]
This issue should probably live under the grommet-site repo, but that isn't apart of the project board yet. Filing here for visibility and prioritization.

- [ ] Confirm root cause of Grommet site failure when deploying static site
- [ ] Assess whether the level of risk for other unexpected failures
- [ ] Make recommendations to harden build process

**Related Issues**
- https://github.com/grommet/grommet-site/issues/197
- https://github.com/grommet/grommet-site/pull/199

## Background Context
Grommet Site appeared to build and deploy successfully on Netlify, however the site was non-interactive when deployed. Static site pages, assets, and routes were all deployed, however all interactive (i.e. Grommet components) were non-operable.

### TLDR;
The static site's index.html build was missing a script pointing to the minified grommet.js. Previously, grommet.js was included and bundled in the vendors~main.js (which contains references to all dependencies). However, during the static site build grommet.js began being chunked into its own, distinct minified file, and no longer bundled and chunked with the other dependencies.
  - EDIT: Potential solutions --> Investigate `webpack.config.optimization.splitChunks` ([Documentation here](https://webpack.js.org/guides/caching/#extracting-boilerplate)) which seems to affect the bundle chunks [applied here](https://github.com/grommet/grommet-site/blob/a56fe07aa68c78674edfa8ba5495be30246fd68d/webpack.config.js#L56).

**Why?** 
The working theory is that the Grommet bundle size surpassed a size threshold where either the build-static-site script or Webpack decides a project should be bundles with other dependencies or on its own.
- [ ] Confirm where this threshold exists and abilities to configure
- [ ] **EDIT:** As I am writing this up, I believe `webpack.config.optimization.splitChunks` might be the primary area to investigate. See "OPPORTUNITY" below.

### Detail
- In the development environment, the site built and ran fine.
- The production deploy built and deployed, but was non-interactional.
- Rolling back to previous commits did not work because the issue was related to the resulting bundle from the build script, not the site code.

#### Deploy Steps
The production Grommet Site is built and deployed as a static site. This is to help with search engine indexing and ranking.
- Netlify runs `yarn build-static` which runs in two steps:
  - `yarn build` which executes webpack and results in the bundle chunks produced.
    - **`grommet-[hash].js` is a new chunk which was not being produced from deploys prior to Jan 8. Previously it had been included with `vendors~main-[hash].js`**.
 
![Screen Shot 2021-01-12 at 3 11 42 PM](https://user-images.githubusercontent.com/1756948/104492329-b4172780-5590-11eb-8173-738ed5ec7e2c.png)

OPPORTUNITY - Investigate `webpack.config.optimization.splitChunks` ([Documentation here](https://webpack.js.org/guides/caching/#extracting-boilerplate)) which seems to affect the bundle chunks [applied here](https://github.com/grommet/grommet-site/blob/a56fe07aa68c78674edfa8ba5495be30246fd68d/webpack.config.js#L56). 

  - `yarn build-pages` which executes `build-static-site.js`
    - This loops through and generates each of the static html files
    - As part of each html file, required scripts are consumed from the Webpack manifest and injected into the html pages. **Because the `grommet-[hash].js` chunk was newly on its own, it was not included in this step and resulted in the site failure**.

## Potential Mitigations to Consider
- [ ] Webpack config or build script adjustments
- [ ] Consider alternate build tooling such as Snowpack (supposed to be simpler, faster, and have an overall good developer experience)
- [ ] E2E smoke test as part of CI/CD (e.g. Cypress, TestCafe) to ensure basic functionality
- [ ] Update README.md to include instructions for the build and deploy process (make it easier for others to follow and assist with CI/CD).
  - [ ] Tangent --> `tools/release-heroku.js` still is in project, but should probably be removed since Netlify is now hosting.
- [ ] Others?
